conda_auth/handlers/base.py

```diff
@@ -0,0 +1,158 @@
+        """Used to retrieve credentials and store them on the ``cache`` property"""
+        if settings.get("auth") == self.get_auth_type():
+            extra_params = {
+                param: settings.get(param) for param in self.get_config_parameters()
+            }
+            self.set_secrets(channel, extra_params)
+
+    @abstractmethod
+    def set_secrets(self, channel: Channel, settings: dict[str, str | None]) -> None:
+        """Implementations should include routine for fetching and storing secrets"""
+
+    @abstractmethod
+    def remove_secrets(self, channel: Channel, settings: dict[str, str | None]) -> None:
+        """Implementations should include routine for removing secrets"""
+
+    @abstractmethod
+    def get_auth_type(self) -> str:
+        """
+        Implementation should return a ``str`` which matches the name of the auth handler type in
+        conda's configuration.
+        """
+
+    @abstractmethod
+    def get_config_parameters(self) -> tuple[str, ...]:
+        """
+        Implementations should return a ``tuple`` of ``str`` which represent the configuration
+        values to use in the ``context.channel_settings`` object.
+        """
+
+    @abstractmethod
+    def get_keyring_id(self, channel_name: str) -> str:
+        """
+        Implementation should return the keyring id that will be used by the manager classes
+        """
+
+
+class CacheChannelAuthBase(ChannelAuthBase):
+    """
+    Adds a class instance cache object for storage of authentication information.
+    """
+
+    def __init__(self, channel_name: str):
+        """
+        Makes sure we have initialized the cache object.
+        """
+        super().__init__(channel_name)
+
+        if not hasattr(self, "_cache"):
+            raise CondaAuthError(
+                "Cache not initialized on class; please run `BasicAuthHandler.set_cache`"
+                " before using"
+            )
+
+    @classmethod
+    def set_cache(cls, cache: dict[str, Any]) -> None:
+        cls._cache = cache
+
+
+def test_credentials(func):
+    """
+    Decorator function used to test whether the collected credentials can successfully make a
+    request.
+
+    This decorator could be applied to any function which updates the ``AuthManager.cache``
+    property.
+    """
+
+    @wraps(func)
+    def wrapper(self, channel: Channel, *args, **kwargs):
+        func(self, channel, *args, **kwargs)
+
+        for url in channel.base_urls:
+            session = get_session(url)
+            resp = session.head(url)
+
+            try:
+                resp.raise_for_status()
+            except requests.exceptions.HTTPError as exc:
+                if exc.response.status_code == requests.codes["unauthorized"]:
+                    error_message = INVALID_CREDENTIALS_ERROR_MESSAGE
+                else:
+                    error_message = str(exc)
+
+                raise CondaAuthError(error_message)
+
```

Comment:
in case the function returns something, it's an easy way to handle that in the decorator

```suggestion
        result = func(self, channel, *args, **kwargs)

        for url in channel.base_urls:
            session = get_session(url)
            resp = session.head(url)

            try:
                resp.raise_for_status()
            except requests.exceptions.HTTPError as exc:
                if exc.response.status_code == requests.codes["unauthorized"]:
                    error_message = INVALID_CREDENTIALS_ERROR_MESSAGE
                else:
                    error_message = str(exc)

                raise CondaAuthError(error_message)

        return result

```

Reply:
Oh, on second thought, shouldn't the calling of the decorated function be happening AFTER the call to the channel URL?

```suggestion
        for url in channel.base_urls:
            session = get_session(url)
            resp = session.head(url)

            try:
                resp.raise_for_status()
            except requests.exceptions.HTTPError as exc:
                if exc.response.status_code == requests.codes["unauthorized"]:
                    error_message = INVALID_CREDENTIALS_ERROR_MESSAGE
                else:
                    error_message = str(exc)

                raise CondaAuthError(error_message)
        return func(self, channel, *args, **kwargs)


```